src/utilities/ci.jl

```diff
@@ -13,12 +14,49 @@ end
 api_hostname(::GitHubActions) = "https://api.github.com"
 clone_hostname(::GitHubActions) = "github.com"
 
-github_repository(env::AbstractDict=ENV) = env["GITHUB_REPOSITORY"]
-github_token(env::AbstractDict=ENV) = env["GITHUB_TOKEN"]
+ci_repository(::GitHubActions, env::AbstractDict=ENV) = env["GITHUB_REPOSITORY"]
+ci_token(::GitHubActions, env::AbstractDict=ENV) = env["GITHUB_TOKEN"]
+
+function get_api_and_repo(ci::GitHubActions, hostname_for_api::AbstractString)
+    token = GitHub.Token(ci_token(ci))
+    api = GitHub.GitHubAPI(; token=token, url=hostname_for_api)
+    repo, _ = @mock GitForge.get_repo(api, ci_repository(ci))
+    return api, repo
+end
+
+struct GitLabCI <: CIService
+    username::String
+    email::String
+end
+
+function GitLabCI()
+    return GitLabCI("gitlab-ci[bot]", "gitlab-ci[bot]@gitlab.com")
+end
+
+api_hostname(::GitLabCI) = "https://gitlab.com/api/v4"
+clone_hostname(::GitLabCI) = "gitlab.com"
+
+ci_repository(::GitLabCI, env::AbstractDict=ENV) = env["CI_PROJECT_PATH"]
+ci_token(::GitLabCI, env::AbstractDict=ENV) = env["GITLAB_TOKEN"]
+
+function get_api_and_repo(ci::GitLabCI, hostname_for_api::AbstractString)
+    token = GitLab.PersonalAccessToken(ci_token(ci))
+    api = GitLab.GitLabAPI(; token=token, url=hostname_for_api)
+    repo, _ = @mock GitForge.get_repo(api, ci_repository(ci))
+    return api, repo
+end
+
+function get_api_and_repo(ci::Any, hostname_for_api::AbstractString)
+    err = "Unknown CI Config: $(typeof(ci))"
+    @error(err)
+    throw(ErrorException(err))
```

Comment:
I don't understand what we're doing here?

Reply:
Printing an error to the log and then throwing an exception. This will only get called if we run `get_api_and_repo` with a `CIService` that doesn't properly implement this function.